### PR TITLE
[DataGrid] Add sort icon for when column is unsorted

### LINKED ISSUE #1076: [DataGrid] Allow sortable indicator on column headers without a sort applied
When a sortable column is not sorted is there any way to surface a 3rd icon type with the existing API? As it stands, it appears the sort icon is removed from the DOM altogether when a column is unsorted.

**3rd Type: Sortable but no sort applied:**
Screen Shot 2021-02-18 at 2 47 47 PM
Ascending:
Screen Shot 2021-02-18 at 2 47 37 PM
Descending:
Screen Shot 2021-02-18 at 2 47 56 PM

- [X] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Summary 💡

Columns that have `sortable:true` but have a 'null' sort direction can display a 3rd neutral sort icon to indicate they are sortable.

## Examples 🌈

In the following example, the table is sorted by Student Name, School has no sort applied but displays a neutral 'unsorted/sortable' indicator.

Screen Shot 2021-02-18 at 2 42 29 PM

## Motivation 🔦

We focus a lot on explicit discoverability in our UIs. Unsorted columns having no indicator that they are sortable does not allow us to give explicit feedback that columns _can_ be sorted and causes users not to use sorting. I'm aware I can apply a default sort with sort model, but this does not guarantee users can discover that _other_ columns are sortable.

### PR DESCRIPTION
**Pull request by m4theushw**
Closes #1076 
Preview: https://deploy-preview-1415--material-ui-x.netlify.app/components/data-grid/demo

This PR changes the usage of the `hideSortIcons` prop. Using `false` will display an intermediary icon when no sort is applied. To keep the old behavior I've added another `unsorted` (default now) value to only show the sort icon when there's a sort applied.

### CODE REVIEW

**Comment by dtassone** [comment]
Why do we need the 'unsorted' value exactly?  

So does `hideSortIcons` allow to hide the sort icon for a specific column?

**Comment by m4theushw** [comment]
> Why do we need the 'unsorted' value exactly?

`unsorted` is to not show the intermediary icon when the column is unsorted. It reproduces the current behavior. Here the age column uses it and the name column is `false`



> So does hideSortIcons allow to hide the sort icon for a specific column?

Yep

File: packages/grid/_modules_/grid/components/containers/GridRootStyles.ts [hunk 1]
```diff
@@ -121,8 +117,18 @@ export const useStyles = makeStyles(
         '& .MuiDataGrid-colCellCenter .MuiDataGrid-colCellTitleContainer': {
           justifyContent: 'center',
         },
-        '& .MuiDataGrid-colCellRight .MuiDataGrid-colCellTitleContainer': {
-          justifyContent: 'flex-end',
+        '& .MuiDataGrid-colCellCenter.MuiDataGrid-withColumnMenu .MuiDataGrid-colCellTitleContainer': {
+          transform: 'translateX(-10px)',
```
Comment (m4theushw) [inline packages/grid/_modules_/grid/components/containers/GridRootStyles.ts:121]: Another option is to add `position: absolute` to the menu so it's not considered as part of the div when aligning to the center. This alternative also requires to apply a background or the header name will go over the column menu icon.

image

**Comment by dtassone** [comment]
`hideSortIcons` I think this `unsorted` setting value is really confusing (to me at least).

To summarise: 
- users can set 3 icons, one for each direction and one when the column is unsorted.
- users might want to hide the sort icons for all columns. For this, he can override the icons and put an empty component, or apply `hideSortIcons` to every sortable columns.
- users might want to hide the sort icons for a specific column. Quite specific edge case IMO, but it could do it using `hideSortIcons`
- users might want to hide the `unsorted` icon for a specific column but show it for another one. `hideSortIcons='unsorted'`. TBH I don't that edge case is necessary. I believe that if users want to hide the unsorted icon, they would do it for all columns, it doesn't make sense to do it for one column and not another. It would be misleading. IMHO, I think we should remove the `unsorted` option for the reason above. Also the prop name is not ideal as it seems like a bool value, but it can take `unsorted`.

**Comment by m4theushw** [comment]
> users might want to hide the sort icons for all columns. For this, he can override the icons and put an empty component, or apply hideSortIcons to every sortable columns.

If he changes the icon to null/false/undefined the default icon is used. If changed to something like `div` the button is rendered without the icon, but still takes space.

> users might want to hide the unsorted icon for a all columns, in that case, it can just put an empty component in the unsorted icon.

Same problem as above.

> users might want to hide the unsorted icon for a specific column but show it for another one. hideSortIcons='unsorted'. TBH I don't that edge case is necessary. I believe that if users want to hide the unsorted icon, they would do it for all columns, it doesn't make sense to do it for one column and not another. It would be misleading. IMHO, I think we should remove the unsorted option for the reason above. Also the prop name is not ideal as it seems like a bool value, but it can take unsorted.

The idea of controlling if the unsorted icon should be displayed came from [this](https://github.com/mui-org/material-ui-x/issues/1076#issuecomment-782903407) comment. However, instead of changing to an enum and creating a breaking change, I kept the boolean but added the third option. The [Autocomplete API](https://next.material-ui.com/api/autocomplete/#props) uses this approach in `blurOnSelect` and `forcePopupIcon`.

File: docs/src/pages/components/data-grid/columns/StylingHeaderGrid.js [hunk 2]
```diff
@@ -7,11 +7,19 @@ const columns = [
     field: 'first',
     headerClassName: 'super-app-theme--header',
     headerAlign: 'center',
+    hideSortIcons: 'unsorted',
   },
   {
     field: 'last',
     headerClassName: 'super-app-theme--header',
     headerAlign: 'center',
+    hideSortIcons: true,
```
Comment (DanailH) [inline docs/src/pages/components/data-grid/columns/StylingHeaderGrid.js:16]: I like the idea but I was thinking of a different API. What if we have a prop like `sortType` and based on its value we choose the icon. So for example possible values can be `dec | asc | unsorted | none`?

**Comment by dtassone** [comment]
- Remove `unsorted` value
- Sortable columns should be identifiable with the unsorted icon
- Icon components should allow null value.

**Comment by oliviertassinari** [comment]
To add a bit to what we have discussed:

- The root pain that #1076 covers is the accessibility **affordance**.
- An new slot for the components prop can solve this issue, it would be rendered for the "unsorted" state on columns that are sortable.
- Is there really a use case for hiding the sort icons? Why not disable the sort on the column in the first place? How about we remove the `hideSortIcons` property? 
- Accepting `null` for a component slot as a way to disable it could be an exception, we have never done this before, we shouldn't generalize it.

File: packages/grid/_modules_/grid/components/columnHeaders/GridColumnHeaderItem.tsx [hunk 3]
```diff
@@ -135,12 +140,11 @@ export const GridColumnHeaderItem = ({
       <GridColumnHeaderSortIcon
         direction={sortDirection}
         index={sortIndex}
-        hide={column.hideSortIcons}
+        hide={column.hideSortIcons || !column.sortable}
```
Comment (dtassone) [inline packages/grid/_modules_/grid/components/columnHeaders/GridColumnHeaderItem.tsx:143]: it doesn't make sense to have an `hide` prop. Might as well do 
```tsx
{column.hideSortIcons || !column.sortable && (
<GridColumnHeaderSortIcon
        direction={sortDirection}
        index={sortIndex}
        />
)}
```

**Comment by dtassone** [comment]
- Issue on this one
image
- Sort icon should be displayed on the right for boolean
image
- Should we show the unsorted icon by default? If so, this icon does not mean unsorted but sorted desc. I would pick either importExport or unfoldMore from the MUI set. We could also apply an hover style which would increase the opacity to make it less intrusive.
- We used to always show the menu icon on top of all icons, as it allows to do any actions, there is a small regression.
now
image
image
image
before
image
image

as a side note, this column menu demo is really bad. It just shows how to disable it
https://deploy-preview-1415--material-ui-x.netlify.app/components/data-grid/columns/#column-menu

**Comment by m4theushw** [comment]
> Issue on this one

Fixed.

> Sort icon should be displayed on the right for boolean
 
Fixed.

> Should we show the unsorted icon by default? If so, this icon does not mean unsorted but sorted desc. I would pick either importExport or unfoldMore from the MUI set. We could also apply an hover style which would increase the opacity to make it less intrusive.

I changed to importExport but I'm still not happy. I think we should only show the icon on hover and use the same icon of the next sort direction. https://github.com/mui-org/material-ui-x/issues/1076#issuecomment-825183525

> We used to always show the menu icon on top of all icons, as it allows to do any actions, there is a small regression.

I related this problem [here](https://github.com/mui-org/material-ui-x/pull/1415#discussion_r613389932). It's a trade-off of fixing the alignment when using `headerAlign=center`. We can add a background but that's a visual breaking change.

**Comment by dtassone** [comment]
> I related this problem here. It's a trade-off of fixing the alignment when using headerAlign=center. We can add a background but that's a visual breaking change.

Yes but the tradeoff seems that it hurts more than the issue. Have you tried with a negative margin?

> I changed to importExport but I'm still not happy. I think we should only show the icon on hover and use the same icon of the next sort direction. #1076 (comment)

Well the issue is to provide the ability to show the unsorted icon when the column is not sorted. If we show the next direction then it would be a different spec. It could be a new option on sortable columns. Something like `showNextDirectionIconOnSortableColumns`, but that would be a different feature.
For the current issue, we have 2 default options.
1. we leave the unsorted icon null.
2. we put an unsorted icon and apply a css to show it on hover. 

As the grid should always come with as many feature as possible, I would lean towards 2.
For css I would go with a low opacity to show it in grey as it's just an info, compared to the active sort state in black.

**Comment by m4theushw** [comment]
> Yes but the tradeoff seems that it hurts more than the issue. Have you tried with a negative margin?

Same problem. I applied a negative `margin-right` to the `.MuiDataGrid-colCellTitleContainer` class of the demo [here](https://material-ui.com/components/data-grid/demo/). Since the column menu is transparent, everything under it will be shown. 

image

> Well the issue is to provide the ability to show the unsorted icon when the column is not sorted. If we show the next direction then it would be a different spec. It could be a new option on sortable columns. Something like showNextDirectionIconOnSortableColumns, but that would be a different feature. Demo candidate in doc? So we show the customizability of the feature or we wait for a user to create the issue.

Yes, for this we added the new icon slot. The problem related in the issue is solved. My point is which icon we would use as the default one. Since we don't have a proper one my idea was to reproduce the same behavior of https://github.com/mui-org/material-ui-x/issues/1076#issuecomment-825183525 here. I just added now the `GridColumnUnsortedIcon` which is a "dynamic icon", showing the up arrow icon or the down arrow icon, depending on the `sortingOrder`. It can be used as the unsorted icon. I'll still have to fix the CSS to increase its alpha when the column is not sorted.

File: docs/pages/api-docs/data-grid.md [hunk 4]
```diff
@@ -114,6 +114,7 @@ Api of the `components` props of type `GridSlotsComponent`
 | <span class="prop-name">ColumnMenuIcon</span> | <span class="prop-type">React.ElementType </span> | <span class="prop-type">TripleDotsVerticalIcon</span> | Icon displayed on the side of the column header title to display the filter input component. |
 | <span class="prop-name">ColumnFilteredIcon</span> | <span class="prop-type">React.ElementType </span> | <span class="prop-type">FilterAltIcon</span> | Icon displayed on the column header menu to show that a filer has been applied to the column. |
 | <span class="prop-name">ColumnSelectorIcon</span> | <span class="prop-type">React.ElementType </span> | <span class="prop-type">ColumnIcon</span> | Icon displayed on the column menu selector tab. |
+| <span class="prop-name">ColumnUnsortedIcon</span> | <span class="prop-type">React.ElementType<br>&#124;&nbsp;null </span> | <span class="prop-type">GridColumnUnsortedIconIcon</span> |  Icon displayed on the side of the column header title when unsorted. |
```
Comment (oliviertassinari) [inline docs/pages/api-docs/data-grid.md:117]:
```suggestion
| <span class="prop-name">ColumnUnsortedIcon</span> | <span class="prop-type">React.ElementType<br>&#124;&nbsp;null </span> | <span class="prop-type">GridColumnUnsortedIcon</span> | Icon displayed on the side of the column header title when unsorted. |
```

File: docs/pages/api-docs/x-grid.md [hunk 5]
```diff
@@ -120,6 +120,7 @@ Api of the `components` props of type `GridSlotsComponent`
 | <span class="prop-name">ColumnMenuIcon</span> | <span class="prop-type">React.ElementType </span> | <span class="prop-type">TripleDotsVerticalIcon</span> | Icon displayed on the side of the column header title to display the filter input component. |
 | <span class="prop-name">ColumnFilteredIcon</span> | <span class="prop-type">React.ElementType </span> | <span class="prop-type">FilterAltIcon</span> | Icon displayed on the column header menu to show that a filer has been applied to the column. |
 | <span class="prop-name">ColumnSelectorIcon</span> | <span class="prop-type">React.ElementType </span> | <span class="prop-type">ColumnIcon</span> | Icon displayed on the column menu selector tab. |
+| <span class="prop-name">ColumnUnsortedIcon</span> | <span class="prop-type">React.ElementType<br>&#124;&nbsp;null </span> | <span class="prop-type">GridColumnUnsortedIconIcon</span> |  Icon displayed on the side of the column header title when unsorted. |
```
Comment (oliviertassinari) [inline docs/pages/api-docs/x-grid.md:123]:
```suggestion
| <span class="prop-name">ColumnUnsortedIcon</span> | <span class="prop-type">React.ElementType<br>&#124;&nbsp;null </span> | <span class="prop-type">GridColumnUnsortedIcon</span> | Icon displayed on the side of the column header title when unsorted. |
```

File: packages/grid/_modules_/grid/components/containers/GridRootStyles.ts [hunk 6]
```diff
@@ -103,26 +103,40 @@ export const useStyles = makeStyles(
...
-        },
-        '& .MuiDataGrid-colCellNumeric .MuiDataGrid-iconButtonContainer': {
-          paddingRight: 5,
+          whiteSpace: 'nowrap',
         },
         '& .MuiDataGrid-colCellSortable': {
           cursor: 'pointer',
+          '&:hover .MuiDataGrid-sortIcon': {
+            opacity: 1,
+          },
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/containers/GridRootStyles.ts:116]: This looks wrong;

- We don't use opacity 1 in the `<Table>` for the hover unsorted state.
- I was also expecting more **isolation** in the style. Why do we specifically target the styles of `GridColumnUnsortedIcon`?

**Review by oliviertassinari — CHANGES_REQUESTED** [review]
- How about we document this in https://deploy-preview-1415--material-ui-x.netlify.app/components/data-grid/sorting/
- It seems that there are broken styles. I don't understand why we use a translateX. This is counterintuitive to me.

**Comment by dtassone** [comment]
As discussed 
1. let's restore the alignment issue
2. fix the opacity

File: packages/grid/_modules_/grid/components/columnHeaders/GridColumnHeaderItem.tsx [hunk 7]
```diff
@@ -140,15 +144,12 @@ export const GridColumnHeaderItem = ({
 
   const columnTitleIconButtons = (
     <React.Fragment>
-      <GridColumnHeaderSortIcon
-        direction={sortDirection}
-        index={sortIndex}
-        hide={column.hideSortIcons}
-      />
       <ColumnHeaderFilterIcon counter={filterItemsCounter} />
+      {column.sortable && !column.hideSortIcons && (
```
Comment (m4theushw) [inline packages/grid/_modules_/grid/components/columnHeaders/GridColumnHeaderItem.tsx:148]: I had to invert the filter icon and the sort icon or it gonna have an empty space when the column is unsorted, but there's an active filter.

image

File: packages/grid/_modules_/grid/components/columnHeaders/GridColumnHeaderSortIcon.tsx [hunk 8]
```diff
@@ -8,51 +8,48 @@ import { GridApiContext } from '../GridApiContext';
...
-  const Icon =
-    direction === 'asc' ? icons!.ColumnSortedAscendingIcon! : icons!.ColumnSortedDescendingIcon!;
-  return <Icon className="MuiDataGrid-sortIcon" />;
+  let Icon = icons!.ColumnUnsortedIcon!;
+  if (direction === 'asc') {
+    Icon = icons!.ColumnSortedAscendingIcon!;
+  } else if (direction === 'desc') {
+    Icon = icons!.ColumnSortedDescendingIcon!;
+  }
+  return <Icon fontSize="small" className="MuiDataGrid-sortIcon" />;
```
Comment (m4theushw) [inline packages/grid/_modules_/grid/components/columnHeaders/GridColumnHeaderSortIcon.tsx:20]: Added `fontSize` to match with `ColumnHeaderFilterIcon`

File: packages/grid/_modules_/grid/components/containers/GridRootStyles.ts [hunk 9]
```diff
@@ -102,27 +102,34 @@ export const useStyles = makeStyles(
...
-        },
-        '& .MuiDataGrid-colCellNumeric .MuiDataGrid-iconButtonContainer': {
-          paddingRight: 5,
+          whiteSpace: 'nowrap',
         },
         '& .MuiDataGrid-colCellSortable': {
           cursor: 'pointer',
         },
-        '& .MuiDataGrid-sortIcon': {
-          fontSize: 18,
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/containers/GridRootStyles.ts:119]: It looks like we should add this style back. The icon should be 18px, not 20px wide.

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/components/containers/GridRootStyles.ts:119]: Well depends on what you read. it feels like 20px in https://material.io/components/data-tables#specs. In https://material-ui.com/components/tables/ we use 18px. On https://search.google.com/, it's 16px:

Screenshot 2021-05-02 at 21 15 21

**Review by oliviertassinari — APPROVED** [review]
Outside of the font size regression?, it looks great.

**Comment by dtassone** [comment]
There are also the issues with the icon on top of the other mentioned 
https://github.com/mui-org/material-ui-x/pull/1415#issuecomment-825681838

**Comment by oliviertassinari** [comment]
@dtassone I don't understand. What's the issue? Could you provide enough information on how to reproduce bugs when you report them?

**Comment by dtassone** [comment]
> > There are also the issues with the icon on top of the other mentioned #1415 (comment)
> 
> @dtassone I don't follow. What's the issue? Could you provide enough information on how to reproduce bugs when you report them?

As mentioned in the link I put, if you reduce the size of the columns then the icons overlap each others
- We used to always show the menu icon on top of all icons, as it allows to do any actions, there is a small regression.
now
image
image
image
before
image
image

**Comment by m4theushw** [comment]
> As mentioned in the link I put, if you reduce the size of the columns then the icons overlap each others

@dtassone That's a side effect of removing `overflow: hidden` from `.MuiDataGrid-colCellTitleContainer`. It was removed because the value in the badge was being cut when reducing the size of the column. I can't think of a way to stack the column menu without adding the background.

With `overflow: hidden`:

Screenshot 2021-02-21 at 19 20 02

**Comment by oliviertassinari** [comment]
> As mentioned in the link I put

The only link I can find is https://deploy-preview-1415--material-ui-x.netlify.app/components/data-grid/columns/#column-menu. I personally can't see any issue.

**Comment by dtassone** [comment]
try overriding the style below.
```tsx
.MuiBadge-anchorOriginTopRightRectangle {
    top: 0;
    right: 0;
    transform: scale(1) translate(30%, -50%);  <= the fix is here, change from 50 to 30%.
    transform-origin: 100% 0%;
}
```

**Comment by m4theushw** [comment]
> try overriding the style below.

@dtassone Same problem. 

image

If we invert the horizontal anchor of the `Badge` we can the solve the regression where the sort icon overlaps the column menu.

image

**Comment by dtassone** [comment]
> > try overriding the style below.
> 
> @dtassone Same problem.

you have to restore the overflow and your changes

File: packages/grid/_modules_/grid/components/containers/GridRootStyles.ts (hunk 9 above)
Comment (m4theushw) [reply packages/grid/_modules_/grid/components/containers/GridRootStyles.ts:119]: I changed to `font-size: inherit` to not have a specific size for just one icon. They appear to have the same dimensions.

image

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/components/containers/GridRootStyles.ts:119]: I see, we rely on IconButton that has font-size 18px on size small. Ok looks good.

**Comment by oliviertassinari** [comment]
> If we invert the horizontal anchor of the Badge we can the solve the regression where the sort icon overlaps the column menu.

It does feel better too. We would only need to make sure it works if the text alignment is reversed. Otherwise, we could reduce the padding of `.MuiDataGrid-colCell` and add some of it on `.MuiDataGrid-colCellTitleContainer`, so that the overflow hidden doesn't hide the bade.

**Review by DanailH — APPROVED** [review]

**Comment by DanailH** [comment]
@m2mathew you would need to accept the new screenshots to fix the build 😉

**Review by dtassone — APPROVED** [review]

**Comment by WilliamZimmermann** [comment]
Hello guys. Finally, the feature to "add sort icon for when column is unsorted" (for <TableSortLabel> component) is not implemented? Because I can't see in the docs (for previous version [4] or the new version [5]) any references to it. If it is not implemented and will not be implemented soon, do you have any suggestion of workaround? Thx!

**Pull request merged**